packages/stream_core/lib/src/utils/shared_emitter.dart

```diff
@@ -2,76 +2,199 @@ import 'dart:async';
 
 import 'package:rxdart/rxdart.dart';
 
-abstract class SharedEmitter<T> {
-  Future<E> waitFor<E extends T>({
-    required Duration timeLimit,
-  });
-
-  StreamSubscription<E> on<E extends T>(void Function(E event) onEvent);
-
+/// A read-only emitter that allows listening for events of type [T].
+///
+/// Listeners can subscribe to receive events, wait for specific event types,
+/// and register handlers for certain event types. The emitter supports
+/// type filtering, allowing listeners to only receive events of a specific
+/// subtype of [T].
+///
+/// See also:
+/// - [MutableSharedEmitter] for the mutable interface that allows emitting events.
+abstract interface class SharedEmitter<T> {
+  /// The stream of events emitted by this emitter.
+  ///
+  /// Returns a [Stream] that emits events of type [T].
+  Stream<T> get stream;
+
+  /// Waits for an event of type [E] to be emitted within the specified [timeLimit].
+  ///
+  /// If such an event is emitted, it is returned. If the time limit
+  /// is exceeded without receiving the event, a [TimeoutException] is thrown.
+  ///
+  /// Returns a [Future] that completes with the first event of type [E].
+  Future<E> waitFor<E extends T>({Duration? timeLimit});
```

Comment:
Maybe we could add an optional `bool Function(E element) test` here, like we used to have in `firstWhere`

Reply:
we already have a dedicated `firstWhere` apart from `waitFor` in the emitter

Reply:
Yeah, but the firstWhere used to have a `required Duration timeLimit,`, that's gone now. So I'm also fine with adding back the timeLimit to firstWhere. 

Now if you want to have a specific filter with a timeout both methods don't quite work.